hal/api/EventAdapter.h

```diff
@@ -0,0 +1,302 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_EVENT_ADAPTER_H
+#define MBED_EVENT_ADAPTER_H
+
+#include "Callback.h"
+#if MBED_CONF_MBED_EVENTS_PRESENT
+#include "Event.h"
+#endif
+#include <stdint.h>
+
+namespace mbed {
+
+#if MBED_CONF_MBED_EVENTS_PRESENT
+
+/** An EventAdapter holds either a Callback<void()> or an Event<void()>. It can be used
+ *  for attach()-style functions to specify if the attached callback should run in user
+ *  context or in interrupt context. To run in user context, construct an EventAdapter
+ *  from an Event object, this will be enqueued in an EventQueue. To run in interrupt
+ *  context (without enqueuing), const an EventAdapter from a Callback object. Example:
+ *
+ *  @code
+ *  InterruptIn sw(SW2);
+ *  EventQueue queue(128);
+ *
+ *  void handler(void) {
+ *  }
+ *
+ *  int main() {
+ *      sw.rise(queue.event(handler)); // run in user context
+ *      sw.rise(callback(handler)); // run in interrupt context
+ *      sw.rise(handler); // shortcut for the form above (run in interrupt context)
+ *  }
+ */
+class EventAdapter {
+
+    typedef Callback<void()> adapter_callback_t;
+    typedef events::Event<void()> adapter_event_t;
```

Comment:
I'm concerned these typedefs don't help much and actually hide information about the type (such as the arguments). They made more sense with the convoluded function-pointer syntax, but with these classes:
``` cpp
EventAdapter(const adapter_callback_t& cb);
// vs 
EventAdapter(const Callback<void()> &cb);
```